File: python/remotetable/backends/ethercalc.py
"""ethercalc backend — socialcalc CSV over HTTP."""

from __future__ import annotations

import csv
import io
import json
import urllib.error
import urllib.request
from typing import Any, List

from .base import Backend


class EtherCalcBackend(Backend):
    """
    Token file JSON optional:
      { "base_url": "https://ethercalc.example", "room": "mysheet", "auth": "optional" }
    Or pass base_url/room in constructor.
    """

    backend_id = "ethercalc"

    def __init__(
        self,
        base_url: str | None = None,
        room: str | None = None,
        token_file: str | None = None,
    ):
        cfg: dict[str, Any] = {}
        if token_file:
            cfg = json.loads(open(token_file, encoding="utf-8").read())
        self.base_url = (base_url or cfg.get("base_url") or "").rstrip("/")
        self.room = room or cfg.get("room") or "sheet"
        self.auth = cfg.get("auth") or cfg.get("access_token") or ""

    def _headers(self) -> dict[str, str]:
        h = {"Content-Type": "text/csv"}
        if self.auth:
            h["Authorization"] = f"Bearer {self.auth}"
        return h

    def _url(self, path: str = "") -> str:
        return f"{self.base_url}/{self.room}{path}"

    def _get(self, path: str = ".csv") -> str:
        req = urllib.request.Request(self._url(path), headers=self._headers(), method="GET")
        with urllib.request.urlopen(req, timeout=60) as resp:
            return resp.read().decode("utf-8", errors="replace")

    def _clear_room_cells(self) -> None:
        """Best-effort wipe so replace is not stacked-append (audreyt/ethercalc POST pastes)."""
        # SocialCalc command via POST /_/{room} JSON
        try:
            body = json.dumps(["set A1:ZZ999 empty"]).encode("utf-8")
            req = urllib.request.Request(
                f"{self.base_url}/_/{self.room}",
                data=body,
                headers={"Content-Type": "application/json"},
                method="POST",
            )
            with urllib.request.urlopen(req, timeout=30) as resp:
                resp.read()
        except Exception:
            pass
        try:
            req = urllib.request.Request(self._url(), method="DELETE")
            with urllib.request.urlopen(req, timeout=15) as resp:
                resp.read()
        except Exception:
            pass

    def _put_csv(self, text: str, *, replace: bool = False) -> None:
        # EtherCalc (common image): POST /_/{room} with text/csv *appends* a paste block.
        # For replace: clear cells first, then POST full CSV grid once.
        if replace:
            self._clear_room_cells()
        data = text.encode("utf-8")
        # Prefer POST /_/{room} (works on audreyt/ethercalc); PUT often 404
        req2 = urllib.request.Request(
            f"{self.base_url}/_/{self.room}",
            data=data,
            headers=self._headers(),
            method="POST",
        )
        try:
            with urllib.request.urlopen(req2, timeout=60) as resp:
                resp.read()
            return
        except urllib.error.HTTPError:
            pass
        req = urllib.request.Request(
            self._url(),
            data=data,
            headers=self._headers(),
            method="PUT",
        )
        with urllib.request.urlopen(req, timeout=60) as resp:
            resp.read()

    def test_connection(self) -> dict[str, Any]:
        if not self.base_url:
            return {"ok": False, "message": "missing base_url", "code": "auth"}
        try:
            self._get(".csv")
            return {"ok": True, "message": f"ethercalc room={self.room}"}
        except Exception as e:
            return {"ok": False, "message": str(e)[:200], "code": "network"}

    def list_tabs(self) -> List[str]:
        # One room ≈ one tab
        return [self.room]

    def ensure_headers(self, tab: str, headers: List[str]) -> dict[str, Any]:
        data = self.read_rows(tab)
        if not data["headers"]:
            self.write_rows(tab, headers, [], mode="replace")
            return {"ok": True, "headers": list(headers)}
        new_h = list(data["headers"])
        for h in headers:
            if h not in new_h:
                new_h.append(h)
        if new_h != data["headers"]:
            self.write_rows(tab, new_h, data["rows"], mode="replace")
        return {"ok": True, "headers": new_h}

    def read_rows(self, tab: str) -> dict[str, Any]:
        text = self._get(".csv")
        reader = csv.reader(io.StringIO(text))
        rows = [list(r) for r in reader]
        # EtherCalc sometimes prefixes a blank/comma-only line before headers.
        while rows and all(not str(c).strip() for c in rows[0]):
            rows = rows[1:]
        if not rows:
            return {"headers": [], "rows": []}
        headers = [str(c) for c in rows[0]]
        body = []
        for r in rows[1:]:
            row = [str(c) for c in r]
            while len(row) < len(headers):
                row.append("")
            body.append(row)
        return {"headers": headers, "rows": body}

    def write_rows(
        self,
        tab: str,
        headers: List[str],
        rows: List[List[str]],
        mode: str = "append",
    ) -> dict[str, Any]:
        if mode == "replace":
            buf = io.StringIO()
            w = csv.writer(buf)
            w.writerow(headers)
            for r in rows:
                w.writerow(r)
            self._put_csv(buf.getvalue(), replace=True)
            return {"written": len(rows)}
        existing = self.read_rows(tab)
        all_rows = existing["rows"] + [list(r) for r in rows]
        hdr = existing["headers"] or list(headers)
        buf = io.StringIO()
        w = csv.writer(buf)
        w.writerow(hdr)
        for r in all_rows:
            w.writerow(r)
        # Full rewrite of combined grid (append semantics without stacking CSV posts)
        self._put_csv(buf.getvalue(), replace=True)
        return {"written": len(rows)}

File: go/remotetable/ethercalc.go
package remotetable

import (
	"encoding/csv"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// EtherCalcBackend talks socialcalc CSV over HTTP.
type EtherCalcBackend struct {
	BaseURL string
	Room    string
	Auth    string
	Client  *http.Client
}

func NewEtherCalc(baseURL, room, auth string) *EtherCalcBackend {
	if room == "" {
		room = "sheet"
	}
	return &EtherCalcBackend{
		BaseURL: strings.TrimRight(baseURL, "/"),
		Room:    room,
		Auth:    auth,
		Client:  &http.Client{Timeout: 60 * time.Second},
	}
}

func (e *EtherCalcBackend) BackendID() string { return BackendEthercalc }

func (e *EtherCalcBackend) headers() http.Header {
	h := make(http.Header)
	h.Set("Content-Type", "text/csv")
	if e.Auth != "" {
		h.Set("Authorization", "Bearer "+e.Auth)
	}
	return h
}

func (e *EtherCalcBackend) roomURL(suffix string) string {
	return e.BaseURL + "/" + e.Room + suffix
}

func (e *EtherCalcBackend) getCSV() (string, error) {
	req, err := http.NewRequest(http.MethodGet, e.roomURL(".csv"), nil)
	if err != nil {
		return "", err
	}
	req.Header = e.headers()
	resp, err := e.Client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body)[:min(200, len(body))])
	}
	return string(body), nil
}

func (e *EtherCalcBackend) putCSV(text string) error {
	req, err := http.NewRequest(http.MethodPut, e.roomURL(""), strings.NewReader(text))
	if err != nil {
		return err
	}
	req.Header = e.headers()
	resp, err := e.Client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		return nil
	}
	// fallback POST append endpoint
	req2, err := http.NewRequest(http.MethodPost, e.BaseURL+"/_/"+e.Room, strings.NewReader(text))
	if err != nil {
		return err
	}
	req2.Header = e.headers()
	resp2, err := e.Client.Do(req2)
	if err != nil {
		return err
	}
	defer resp2.Body.Close()
	if resp2.StatusCode < 200 || resp2.StatusCode >= 300 {
		b, _ := io.ReadAll(resp2.Body)
		return fmt.Errorf("HTTP %d: %s", resp2.StatusCode, string(b)[:min(200, len(b))])
	}
	return nil
}

func parseCSVGrid(text string) (headers []string, rows [][]string, err error) {
	r := csv.NewReader(strings.NewReader(text))
	r.FieldsPerRecord = -1
	all, err := r.ReadAll()
	if err != nil {
		return nil, nil, err
	}
	// Skip blank / comma-only prefix rows (EtherCalc quirk).
	for len(all) > 0 {
		blank := true
		for _, c := range all[0] {
			if strings.TrimSpace(c) != "" {
				blank = false
				break
			}
		}
		if !blank {
			break
		}
		all = all[1:]
	}
	if len(all) == 0 {
		return nil, nil, nil
	}
	headers = all[0]
	if len(all) > 1 {
		rows = all[1:]
	}
	return headers, rows, nil
}

func encodeCSVGrid(headers []string, rows [][]string) string {
	var b strings.Builder
	w := csv.NewWriter(&b)
	_ = w.Write(headers)
	for _, row := range rows {
		_ = w.Write(row)
	}
	w.Flush()
	return b.String()
}

func (e *EtherCalcBackend) TestConnection() (bool, string) {
	if e.BaseURL == "" {
		return false, "missing base_url"
	}
	if _, err := e.getCSV(); err != nil {
		return false, err.Error()
	}
	return true, "ethercalc room=" + e.Room
}

func (e *EtherCalcBackend) ListTabs() []string {
	// Single-room model: room name is the only "tab".
	return []string{e.Room}
}

func (e *EtherCalcBackend) EnsureHeaders(tab string, headers []string) []string {
	curH, curRows := e.ReadRows(tab)
	if len(curH) == 0 {
		_ = e.putCSV(encodeCSVGrid(headers, nil))
		return headers
	}
	merged := append([]string{}, curH...)
	for _, h := range headers {
		found := false
		for _, c := range merged {
			if c == h {
				found = true
				break
			}
		}
		if !found {
			merged = append(merged, h)
			for i := range curRows {
				curRows[i] = append(curRows[i], "")
			}
		}
	}
	if len(merged) != len(curH) {
		_ = e.putCSV(encodeCSVGrid(merged, curRows))
	}
	return merged
}

func (e *EtherCalcBackend) ReadRows(tab string) ([]string, [][]string) {
	text, err := e.getCSV()
	if err != nil {
		return nil, nil
	}
	h, rows, err := parseCSVGrid(text)
	if err != nil {
		return nil, nil
	}
	return h, rows
}

func (e *EtherCalcBackend) WriteRows(tab string, headers []string, rows [][]string, mode string) int {
	e.EnsureHeaders(tab, headers)
	curH, curRows := e.ReadRows(tab)
	if len(curH) == 0 {
		curH = headers
	}
	// pad
	padded := make([][]string, len(rows))
	for i, r := range rows {
		nr := make([]string, len(curH))
		copy(nr, r)
		padded[i] = nr
	}
	out := padded
	if mode != "replace" {
		out = append(curRows, padded...)
	}
	_ = e.putCSV(encodeCSVGrid(curH, out))
	return len(padded)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
